### PR TITLE
FE-808 | Add TS types for streaming

### PR DESCRIPTION
**Pull request by ryancharris**
### Notes
[Jira Ticket](https://faunadb.atlassian.net/browse/FE-808)

Eric Wyne reported that our TS types were not updated when adding streaming support. This PR is intended to do that 👍

### CODE REVIEW

**Comment by ryancharris** [comment]
@BrunoQuaresma @rodrimaia I was looking at the TS docs and it seems we can auto-generate our `.d.ts` files by updating our `.tsconfig` file: https://www.typescriptlang.org/docs/handbook/declaration-files/dts-from-js.html

What do you think?

File: src/types/Stream.d.ts [hunk 1]
```diff
@@ -1,10 +1,27 @@
...
+
+export interface StreamClient {
+  client: Client
+  expression: ExprArg
+  options: Options
+  onEvent: Function
+}
+
+export interface Options {
+  ref: string
```
Comment (BrunoQuaresma) [inline src/types/Stream.d.ts:21]: Is this a string or a Ref object?

File: src/types/Stream.d.ts [hunk 2]
```diff
@@ -1,10 +1,27 @@
...
 
 export interface EventDispatcher {
   allowedEvents: string[]
 }
+
+export interface StreamClient {
+  client: Client
+  expression: ExprArg
+  options: Options
+  onEvent: Function
```
Comment (BrunoQuaresma) [inline src/types/Stream.d.ts:17]: This is too generic, what should we expect here? what arguments and response?

File: src/types/Stream.d.ts [hunk 3]
```diff
@@ -1,10 +1,27 @@
...
+import { ExprArg } from './query'
 
 export interface Subscription {
-  client: Client
+  client: StreamClient
   dispatcher: EventDispatcher
 }
 
 export interface EventDispatcher {
   allowedEvents: string[]
```
Comment (BrunoQuaresma) [inline src/types/Stream.d.ts:10]: This looks too generic as well, we should say specifically the events.

File: src/types/Stream.d.ts [hunk 4]
```diff
@@ -1,10 +1,27 @@
...
+  onEvent: Function
+}
+
+export interface Options {
+  ref: string
+  ts: string
+  new: string
+  old: string
+  diff: string
+  action: string
```
Comment (BrunoQuaresma) [inline src/types/Stream.d.ts:26]: I also think this is too generic. If the action is pre-defined, we should put the specific values.

**Review by BrunoQuaresma — COMMENTED** [review]
Some notes.

File: src/types/Stream.d.ts (hunk 2 above)
Comment (ecwyne) [reply src/types/Stream.d.ts:17]: This may help https://cjam.github.io/typesafe-event-emitter-in-typescript/

File: src/types/Client.d.ts [hunk 5]
```diff
@@ -13,6 +14,7 @@ export interface ClientConfig {
   keepAlive?: boolean
   headers?: { [key: string]: string | number }
   fetch?: typeof fetch
+  stream?: Subscription
```
Comment (erickpintor) [inline src/types/Client.d.ts:17]: I think this filed is misplaced. There isn't a `stream` field in the client config object as fas as I know.

File: src/types/Client.d.ts [hunk 6]
```diff
@@ -25,4 +27,5 @@ export default class Client {
   query<T = object>(expr: ExprArg, options?: QueryOptions): Promise<T>
   paginate(expr: Expr, params?: object, options?: QueryOptions): PageHelper
   ping(scope?: string, timeout?: number): Promise<string>
+  stream(expr: Expr, options?: object): Subscription
```
Comment (erickpintor) [inline src/types/Client.d.ts:30]: Could you also type this `options` argument? It has a single field called `fields` and it is a list of strings.

File: src/types/Stream.d.ts (hunk 3 above)
Comment (erickpintor) [reply src/types/Stream.d.ts:10]: I don't think it's worth documenting the internal types. Users don't have access to it.

File: src/types/Stream.d.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import Client from './Client'
+import { ExprArg } from './query'
+
+export interface Subscription {
+  client: StreamClient
+  dispatcher: EventDispatcher
```
Comment (erickpintor) [inline src/types/Stream.d.ts:6]: This fields don't exist, their `_` prefixed counterparts are private to the subscription class.

Comment (erickpintor) [reply src/types/Stream.d.ts:6]: I think we should document its public interface instead: `on`, `start`, and `close`. Also, the most useful type annotation will probably by the events themselves.

File: src/types/Stream.d.ts (hunk 2 above)
Comment (erickpintor) [reply src/types/Stream.d.ts:17]: This class is internal. We should focus on the public interface, IMO.

File: src/types/Stream.d.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+export interface StreamClient {
+  client: Client
+  expression: ExprArg
+  options: Options
+  onEvent: Function
+}
+
+export interface Options {
```
Comment (erickpintor) [inline src/types/Stream.d.ts:20]: What is this object for?

**Review by erickpintor — CHANGES_REQUESTED** [review]

File: src/types/Stream.d.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import Client from './Client'
+import { ExprArg } from './query'
+
+export interface Subscription {
+  on: (type: SubscriptionEvents, callback: Function) => this
+  start: () => this
+  close: () => this
+}
+
+type SubscriptionEvents = 'ref' | 'ts' | 'new' | 'old' | 'diff' | 'action'
```
Comment (erickpintor) [inline src/types/Stream.d.ts:10]: These are fields that come with the "version" event. The event types are: "start", "version", "history_rewrite", and "error". The list of events can be found in the docs here: https://github.com/fauna/faunadb-js/blob/master/src/stream.js#L289-L372

Here is the full list of events and their fields for convenience:

```
event: start
  fields:
    - event: the event name
    - txnTS: the event transaction timestamp
    - data: the stream start timestamp

event: version
  fields:
    - event: the event name
    - txnTS: the event transaction timestamp
    - data:
      - ref: the document's ref
      - ts: the document's valid timestamp
      - action: the action that triggered the event
      - new: the document's new data
      - old: the document's old data (optional)
      - diff: the diff between old and new data (optional)

event: history_rewrite
  fields:
    - event: the event name
    - txnTS: the event transaction timestamp
    - data:
      - ref: the document's ref
      - ts: the document's valid timestamp
      - action: the action that triggered the event

event: error
  fields:
    - event: the event name
    - txnTS: the event transaction timestamp
    - data:
      - code: the error code
      - description: the error description
```

File: src/types/Client.d.ts [hunk 10]
```diff
@@ -25,4 +26,5 @@ export default class Client {
   query<T = object>(expr: ExprArg, options?: QueryOptions): Promise<T>
   paginate(expr: Expr, params?: object, options?: QueryOptions): PageHelper
   ping(scope?: string, timeout?: number): Promise<string>
+  stream(expr: Expr, options?: { fields: string[] }): Subscription
```
Comment (erickpintor) [inline src/types/Client.d.ts:29]: `fields` itself is also optional.

File: src/types/Stream.d.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import Client from './Client'
+import { ExprArg } from './query'
+import { FaunaTime, Ref } from './values'
+
+export interface Subscription {
+  on: (type: SubscriptionEvents, callback: Function) => this
+  start: () => this
+  close: () => this
```
Comment (erickpintor) [inline src/types/Stream.d.ts:8]: `close` doesn't return anything.

File: src/types/Stream.d.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+type StreamStartEventData = FaunaTime
+
+type StreamVersionEventData = {
+  ref: Ref
+  ts: number
+  action: string
+  new: object
+  old: object
+  diff: string
```
Comment (erickpintor) [inline src/types/Stream.d.ts:26]: `diff` is also an object.

File: src/types/Stream.d.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+type StreamErrorEventData = {
+  code: number
```
Comment (erickpintor) [inline src/types/Stream.d.ts:30]: `code` is a string.

File: src/types/Stream.d.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+  description: string
+}
+
+type StreamHistoryRewriteEventData = {
+  ref: Ref
+  ts: number
+  action: string
+}
+
+type SubscriptionEvents = 'start' | 'version' | 'history_rewrite' | 'error'
```
Comment (erickpintor) [inline src/types/Stream.d.ts:40]: Question: is it possible in TS to tie the literal string with the event type received?

Comment (ecwyne) [reply src/types/Stream.d.ts:40]: Found this. https://github.com/bterlson/strict-event-emitter-types

File: src/types/Stream.d.ts [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+  | StreamStartSubscription
+  | StreamVersionSubscription
+  | StreamErrorSubscription
+  | StreamHistoryRewriteSubscription
+
+interface Subscription {
+  start: () => this
+  close: () => void
+  txnTS: number
+}
```
Comment (erickpintor) [inline src/types/Stream.d.ts:15]: The subscription itself does not contain data. The previews version had the correct type signature:

```js
interface Subscription {
  on: (type: SubscriptionEvents, callback: Function) => this
  start: () => this
  close: () => this
}
```

What I wanted to try is to tie the `SubscriptionEvents` literal string with the type of the `callback` parameter.

For example:

```js
client.stream(ref).on('start', ts => ...)
//                             ^ 
// I wanted typescript to understand that this is a FaunaTime
```

@BrunoQuaresma mentioned earlier in the project that maybe that's possible. I don't know much about TS, though. I'm okay if that's not feasible.

**Comment by ecwyne** [comment]
I believe you're looking for something like this.

```ts
// Fill out these handler type signatures
type SubscriptionEventHandlers = {
    start: (ts: FaunaTime, e: {event: 'start', txnTs: number, data: FaunaTime}) => void;
    error: () => void;
    version: () => void;
    history_rewrite: () => void;
    snapshot: () => void;
};

interface Subscription {
    on: <T extends keyof SubscriptionEventHandlers>(
        type: T,
        callback: SubscriptionEventHandlers[T],
    ) => this;
    start: () => this;
    close: () => void;
}
```

File: src/types/Stream.d.ts (hunk 15 above)
Comment (ecwyne) [reply src/types/Stream.d.ts:15]: https://github.com/fauna/faunadb-js/pull/356#issuecomment-713202337

**Comment by ryancharris** [comment]
@erickpintor What fields am I missing here? https://github.com/fauna/faunadb-js/pull/356/files#diff-21ff9a7035770e3d38455382a3b21b2720ee2fc5894cd9ee15d2502201b2f92cR66

File: src/types/Client.d.ts [hunk 16]
```diff
@@ -25,4 +26,5 @@ export default class Client {
   query<T = object>(expr: ExprArg, options?: QueryOptions): Promise<T>
   paginate(expr: Expr, params?: object, options?: QueryOptions): PageHelper
   ping(scope?: string, timeout?: number): Promise<string>
+  stream(expr: Expr, options?: { fields?: string[] }): Subscription
```
Comment (erickpintor) [inline src/types/Client.d.ts:29]: Not a requirement. More of a question: is it possible to specify the valid values for this array?

File: src/types/Stream.d.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+  on: <T extends keyof SubscriptionEventHandlers>(
+    type: T,
+    callback: SubscriptionEventHandlers[T]
+  ) => this
+  start: () => this
+  close: () => void
+}
+
+type SubscriptionEventHandlers = {
+  start: (type: 'start', txn: values.FaunaTime, event: values.FaunaTime) => void
```
Comment (erickpintor) [inline src/types/Stream.d.ts:15]: Both txn and event are no longer `FaunaTime` instances. They are numbers now.

Comment (erickpintor) [reply src/types/Stream.d.ts:15]: The same thing goes for for all `txn` and `ts` fields in this type annotation.

File: src/types/Stream.d.ts [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+        ref: values.Ref
+        ts: values.FaunaTime
+        data: object
+      }
+      prev: {
+        ref: values.Ref
+        ts: values.FaunaTime
+        data?: object
+      }
+    }
```
Comment (erickpintor) [inline src/types/Stream.d.ts:44]: I appreciate the careful typing here but, I think this are better off as just `object` as their rendering are highly dynamic and dependent on the document being streamed. For instance, data is never guaranteed to be present. Once we enable schema documents, they will have other properties, like `name`, `roles`, etc. Trying to cover all possibilities at this level will be quite tedious and error prone.

Comment (erickpintor) [reply src/types/Stream.d.ts:44]: This is the same for other `document` fields in this type annotation.

Comment (ecwyne) [reply src/types/Stream.d.ts:44]: If I can provide some user feedback. I would much rather have incomplete/error-prone types over simply `any` or `object`. TS types provide not only type-safety for refactoring but also serve as a main source for documentation when learning/exploring new parts of the api.

Comment (erickpintor) [reply src/types/Stream.d.ts:44]: As a user, can you work around incomplete types in TS? If so, I'd say mark data as optional and leave it as is.

**Comment by ryancharris** [comment]
@erickpintor Any idea why this might be passing in one Preview environment, but not the other?

File: src/stream.js [hunk 19]
```diff
@@ -327,9 +327,9 @@ EventDispatcher.prototype.dispatch = function(event) {
  * @type {object}
  * @property {string} event='history_rewrite'
```
Comment (erickpintor) [inline src/stream.js:328]:
```suggestion
 * @property {string} type='history_rewrite'
```

File: src/stream.js [hunk 20]
```diff
@@ -340,9 +340,9 @@ EventDispatcher.prototype.dispatch = function(event) {
  *
  * @event module:stream~Subscription#snapshot
  * @type {object}
- * @property {string} event='version'
+ * @property {string} event='snapshot'
```
Comment (erickpintor) [inline src/stream.js:343]:
```suggestion
 * @property {string} type='snapshot'
```

File: src/stream.js [hunk 21]
```diff
@@ -340,9 +340,9 @@ EventDispatcher.prototype.dispatch = function(event) {
...
  *   The event type.
- * @property {number} txnTS
+ * @property {number} txn
  *   The event's transaction timestamp.
  * @property {object} data
```
Comment (erickpintor) [inline src/stream.js:347]:
```suggestion
 * @property {object} event
```

File: src/stream.js [hunk 22]
```diff
@@ -356,23 +356,21 @@ EventDispatcher.prototype.dispatch = function(event) {
  * @type {object}
  * @property {string} event='error'
```
Comment (erickpintor) [inline src/stream.js:357]:
```suggestion
 * @property {string} type='error'
```

File: src/stream.js [hunk 23]
```diff
@@ -356,23 +356,21 @@ EventDispatcher.prototype.dispatch = function(event) {
...
- *   * 'ref': The event's document reference;
- *   * 'ts': The event's valid time;
- *   * 'new': The event's new data;
- *   * 'old': The event's old data;
- *   * 'diff': The event's data difference;
- *   * 'action': The event's action.
+ *   * 'action': The action type
+ *   * 'document': The document's data
+ *   * 'diff': The difference between 'document' and 'prev'
+ *   * 'prev': The event's old data
```
Comment (erickpintor) [inline src/stream.js:373]:
```suggestion
 *   * 'prev': The event's previews data
```

File: src/types/Stream.d.ts [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+    event: {
+      action: 'history_rewrite'
+      document: object
+    }
+  ) => void
+  snapshot: (
+    type: 'snapshot',
+    txn: number,
+    event: {
+      action: 'snapshot'
```
Comment (erickpintor) [inline src/types/Stream.d.ts:46]: A snapshot has no action. The event itself is the document.

**Comment by erickpintor** [comment]
> Any idea why this might be passing in one Preview environment, but not the other?

We gotta check case by case. I think the node version has something to do with it:

```
TypeError: results.flat is not a function

      49 |   }
      50 | 
    > 51 |   return results.flat()
```

Also, preview has the new event format, while I don't think stable has it yet.

File: src/stream.js (hunk 23 above)
Comment (ryancharris) [reply src/stream.js:373]: @erickpintor I'm a bit confused by this annotation -- I thought the `prev` field represented the old document data?

File: src/types/Stream.d.ts (hunk 24 above)
Comment (ryancharris) [reply src/types/Stream.d.ts:46]: @erickpintor In that case, would `event: object` be the better definition? Or is there an `event.document` field that should be an `object`?

**Comment by ryancharris** [comment]
> Also, preview has the new event format, while I don't think stable has it yet.

@erickpintor What implications will this have?

**Comment by ryancharris** [comment]
> We gotta check case by case. I think the node version has something to do with it:

@erickpintor You are correct, sir! It seems the tests pass whenever we're using Node v11+. Locally (using NVM), I ran the tests with these various Node versions:

```
Tests fail with these Node versions
- 10.18.0
- 10.23.0

Tests pass with these Node versions
- 11.0.0
- 11.15.0
- 12.19.0
- 14.2.0
```

**Comment by ryancharris** [comment]
@erickpintor This is the TLDR of why the tests are passing/failing:

```
** Our environments are using the following Node versions:
These all fail due to Node version issues
- core-stable-10: Node 10
- core-nightly-10: Node 10
- core-preview-10: Node 10

These fail due to feature functionality
- core-stable-11: Node 11
- core-nightly-11: Node 11

This passes since it's the right Node version and has streaming enabled
- core-preview-11: Node 11
```

**Comment by erickpintor** [comment]
Here are my findings:

* The JS driver has code that is broken in node 10 (we gotta fix that);
* Stable docker image has no stream api enabled, hence stream tests will fail with it;
* The night build has the new event’s format, where the branch doesn’t. Therefore, it will fail with it;
* The preview docker image has the old event’s format, which matches with the branch’s code so, that’s why it’s the only scenario passing.

File: src/types/Stream.d.ts (hunk 24 above)
Comment (erickpintor) [reply src/types/Stream.d.ts:46]: `event: object` should it. There's no `document` field in it.

File: src/stream.js (hunk 23 above)
Comment (erickpintor) [reply src/stream.js:373]: It is the old documents data. I thought using the word preview would be more consistent with the field named `prev`. I'm fine with "old" too.

File: test/stream.test.js [hunk 25]
```diff
@@ -92,7 +92,7 @@ describe('StreamAPI', () => {
           client.query(q.Update(doc.ref, {}))
         })
         .on('version', (_, event) => {
-          expect(client.getLastTxnTime()).toEqual(event.txnTS)
+          expect(client.getLastTxnTime()).toEqual(event.event.document.ts)
```
Comment (erickpintor) [inline test/stream.test.js:95]: The transaction time should be tracked against the `event.txn` field.
```suggestion
          expect(client.getLastTxnTime()).toEqual(event.txn)
```

File: src/stream.js (hunk 23 above)
Comment (ryancharris) [reply src/stream.js:373]: @erickpintor Ah, okay. I'll go with `previous` cause that sounds more semantic 👍

File: test/stream.test.js [hunk 26]
```diff
@@ -235,7 +234,7 @@ describe('StreamAPI', () => {
         .start()
     })
 
-    test('filter buffered events prior to snapshot', done => {
+    test.skip('filter buffered events prior to snapshot', done => {
```
Comment (erickpintor) [inline test/stream.test.js:237]: This patch fixes the test:

```diff
diff --git a/src/stream.js b/src/stream.js
index 00d4e36..f5828d9 100644
--- a/src/stream.js
+++ b/src/stream.js
@@ -475,7 +475,7 @@ function StreamAPI(client) {
       dispatcher.dispatch(snapshotEvent)
       for (var i = 0; i < buffer.length; i++) {
         var bufferedEvent = buffer[i]
-        if (bufferedEvent.txnTS > snapshotEvent.data.ts) {
+        if (bufferedEvent.txn > snapshotEvent.event.ts) {
           dispatcher.dispatch(bufferedEvent)
         }
       }
```

File: test/stream.test.js (hunk 25 above)
Comment (ryancharris) [reply test/stream.test.js:95]: @erickpintor That's what I originally thought, but the test wasn't consistently passing 🤔

File: test/stream.test.js (hunk 26 above)
Comment (ryancharris) [reply test/stream.test.js:237]: @erickpintor Of course! I don't know how I missed that 😅 I was looking at this file for way too long.

**Review by erickpintor — APPROVED** [review]
LGTM

File: test/stream.test.js [hunk 27]
```diff
@@ -59,10 +59,10 @@ describe('StreamAPI', () => {
     test('can listen to events', done => {
       stream = client
         .stream(doc.ref)
-        .on('start', (ts, event) => {
-          expect(event.event).toEqual('start')
-          expect(event.txnTS).toBeDefined()
-          expect(event.data).toEqual(ts)
+        .on('start', (_, event) => {
```
Comment (danieltodonnell) [inline test/stream.test.js:62]: I see the _ was the part you were stuck on to get the test running. nice.

**Review by danieltodonnell — APPROVED** [review]
Didn't run it but code LGTM

**Review by nathaniel-may — APPROVED** [review]

**Pull request merged**